gms/gossiper.cc

```diff
@@ -2088,22 +2092,20 @@ future<> gossiper::add_saved_endpoint(inet_address ep) {
     if (es) {
         ep_state = *es;
         logger.debug("not replacing a previous ep_state for {}, but reusing it: {}", ep, ep_state);
-        ep_state.set_heart_beat_state_and_update_timestamp(heart_beat_state());
+        ep_state.update_timestamp();
```

Comment:
We should avoid using non-zero generation and inject gossip sates. The code before this change guarantees it. We can check if the current ep_state already has the dc / rack / tokens /host id, skip injecting if so.  Or, we can reuse the existing  ep_state but reset the generation to zero. It is dangerous to reuse non-zero generation gossip state + injecting. 

Are we calling add_saved_endpoint when we have collected all the states loaded_state and only call it once during startup? We should, this will simplify how we inject.

Reply:
> We should avoid using non-zero generation and inject gossip sates. The code before this change guarantees it.

How?

Please see latest version of https://github.com/scylladb/scylladb/pull/16432. It won't update an existing ep_state if it has non-zero generation.

Reply:
Because `ep_state.set_heart_beat_state_and_update_timestamp(heart_beat_state())` always set the generation to zero.

Reply:
It's better to skip loading local state if there is an existing endpoint state with non-zero generation.

Reply:
The reason is that resetting the ep_state generation to 0 will cause any update from remote with non-zero generation to win over the local endpoint state and replace it anyhow, so a non-zero generation means we should lose the race here too and not bother with updating the endpoint state..